File: mac/tools/crosscheck/overlay.py
# The repository root is on the path already, put there by cross_check.py.
#
# The GEOMETRY and the WORDS are checked here. The drawing is not, and
# `fits` is deliberately not: the two renderers do not draw the same width,
# so agreeing with the other platform's number would mean being wrong about
# this machine's own picture. See docs/MAC-VIDEO-PLAN.md section 3a.
import time
from dropdeck import constants as C
from dropdeck import overlay as O

# The clock is pinned so both copies see the same minute.
time.strftime = lambda fmt, *a: "09:41" if fmt == C.OVERLAY_CLOCK_FORMAT else ""

out = []
for spot in O.PLACES:
    out.append("place|%s|%s|%s|%s" % (spot.key, spot.label, spot.align,
                                      spot.describe_where()))
    for (w, h) in ((1280, 720), (1920, 1080), (854, 480), (640, 360), (3840, 2160)):
        out.append("  rect|%s|%d|%d|%s|%d"
                   % (spot.key, w, h, spot.rect(w, h), spot.text_size(h)))
for key in ("top", "corner", "lower", "clock", "nowhere"):
    out.append("label|%s|%s" % (key, O.place_label(key)))

# describe(), which is what Command Shift V reads out.
CASES = [
    ("nothing", {}),
    ("station-only", {"text_top": "station", "name": "Blindside Radio"}),
    ("station-no-name", {"text_top": "station"}),
    ("two", {"text_top": "station", "name": "Tony's Tunes",
             "text_clock": "time"}),
    ("three", {"text_top": "station", "name": "Tony's Tunes",
               "text_clock": "time", "text_lower": "playing"}),
    ("all-four", {"text_top": "station", "name": "Tony's Tunes",
                  "text_corner": "words", "text_corner_words": "LIVE",
                  "text_lower": "playing", "text_clock": "time"}),
    ("playing-empty", {"text_lower": "playing"}),
    ("words-empty", {"text_corner": "words"}),
    ("file-missing", {"text_lower": "file", "text_lower_file": "/tmp/nope-xyz.txt"}),
    ("stream-name", {"text_top": "station", "stream_name": "Fallback FM"}),
    ("unknown-kind", {"text_top": "weather"}),
]
for name, settings in CASES:
    for title in ("", "Fleetwood Mac - Dreams"):
        ov = O.Overlay(settings)
        ov.set_title(title)
        out.append("describe|%s|%r|%s" % (name, title, ov.describe()))
        out.append("  anything|%s|%s" % (name, ov.anything_on()))
        for key in C.PLACES_ORDER:
            out.append("  kind|%s|%s|%s" % (name, key, ov.kind_of(key)))
print("\n".join(out))

File: mac/tools/crosscheck/overlay.swift
import Foundation

var out: [String] = []
func tup(_ r: (left: Int, top: Int, right: Int, bottom: Int)) -> String {
    "(\(r.left), \(r.top), \(r.right), \(r.bottom))"
}
for spot in Overlays.places {
    out.append("place|\(spot.key)|\(spot.label)|"
             + "\(spot.align == .right ? "right" : "left")|\(spot.describeWhere())")
    for size in [(1280, 720), (1920, 1080), (854, 480), (640, 360), (3840, 2160)] {
        out.append("  rect|\(spot.key)|\(size.0)|\(size.1)|"
                 + "\(tup(spot.rect(size.0, size.1)))|\(spot.textSize(size.1))")
    }
}
for key in ["top", "corner", "lower", "clock", "nowhere"] {
    out.append("label|\(key)|\(Overlays.placeLabel(key))")
}

// The clock is pinned so both copies see the same minute.
let pinned: Date = {
    var c = DateComponents()
    c.year = 2026; c.month = 9; c.day = 8; c.hour = 9; c.minute = 41
    return Calendar(identifier: .gregorian).date(from: c)!
}()

let cases: [(String, OverlaySettings)] = [
    ("nothing", OverlaySettings()),
    ("station-only", { var s = OverlaySettings()
        s.places = ["top": ["kind": "station"]]; s.name = "Blindside Radio"; return s }()),
    ("station-no-name", { var s = OverlaySettings()
        s.places = ["top": ["kind": "station"]]; return s }()),
    ("two", { var s = OverlaySettings()
        s.places = ["top": ["kind": "station"], "clock": ["kind": "time"]]
        s.name = "Tony's Tunes"; return s }()),
    ("three", { var s = OverlaySettings()
        s.places = ["top": ["kind": "station"], "clock": ["kind": "time"],
                    "lower": ["kind": "playing"]]
        s.name = "Tony's Tunes"; return s }()),
    ("all-four", { var s = OverlaySettings()
        s.places = ["top": ["kind": "station"],
                    "corner": ["kind": "words", "words": "LIVE"],
                    "lower": ["kind": "playing"], "clock": ["kind": "time"]]
        s.name = "Tony's Tunes"; return s }()),
    ("playing-empty", { var s = OverlaySettings()
        s.places = ["lower": ["kind": "playing"]]; return s }()),
    ("words-empty", { var s = OverlaySettings()
        s.places = ["corner": ["kind": "words"]]; return s }()),
    ("file-missing", { var s = OverlaySettings()
        s.places = ["lower": ["kind": "file", "file": "/tmp/nope-xyz.txt"]]; return s }()),
    ("stream-name", { var s = OverlaySettings()
        s.places = ["top": ["kind": "station"]]; s.streamName = "Fallback FM"; return s }()),
    ("unknown-kind", { var s = OverlaySettings()
        s.places = ["top": ["kind": "weather"]]; return s }()),
]
for (name, settings) in cases {
    for title in ["", "Fleetwood Mac - Dreams"] {
        let ov = Overlay(settings: settings)
        ov.clock = { pinned }
        ov.setTitle(title)
        out.append("describe|\(name)|'\(title)'|\(ov.describe())")
        out.append("  anything|\(name)|\(ov.anythingOn() ? "True" : "False")")
        for key in C.placesOrder {
            out.append("  kind|\(name)|\(key)|\(ov.kindOf(key))")
        }
    }
}
print(out.joined(separator: "\n"))
